vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -29,6 +72,24 @@ struct ccb
   std::vector<std::string>* original_interactions;
   std::vector<CCB::label> stored_labels;
   size_t action_with_label;
+
+  // In order to reuse the v_arrays in this reduction we are using a combination of object pooling
+  // and tracking queues. Since the v_arrays are used by value we need to shallow copy them into
+  // place. Therefore, we can't use the pooled object directly but must use the buffer within.
+  // 1. Remove object from pool
+  // 2. Shallow copy into location to be used
+  // 3. Place onto corresponding queue When done:
+  // 1. Shallow copy onto an item from queue
+  // 2. Return pooled queue item to pool
+  //
+  // It doesn't matter which queue object is used to return it as it is essentially just a container
+  // for the internal buffer that we are using.
+  std::queue<v_array<CB::cb_class>*> in_use_cb_label;
```

Comment:
I don't understand a fundamental reason why we need in_use_* here.

Reply:
Explained in the previous comment but, all objects returned to the pool must have originated from the pool. This tracks those objects so we can return them.